usr/src/uts/common/io/vioif/vioif.c

```diff
@@ -709,113 +708,104 @@ vioif_unicst(void *arg, const uint8_t *macaddr)
 }
 
 
-static int
+static uint_t
 vioif_add_rx(struct vioif_softc *sc, int kmflag)
 {
-	struct vq_entry *ve;
-	struct vioif_rx_buf *buf;
+	uint_t num_added = 0;
 
-	ve = vq_alloc_entry(sc->sc_rx_vq);
-	if (!ve) {
-		/*
-		 * Out of free descriptors - ring already full.
-		 * It would be better to update sc_norxdescavail
-		 * but MAC does not ask for this info, hence we
-		 * update sc_norecvbuf.
-		 */
-		sc->sc_norecvbuf++;
-		goto exit_vq;
-	}
-	buf = sc->sc_rxbufs[ve->qe_index];
+	for (;;) {
+		struct vq_entry *ve;
+		struct vioif_rx_buf *buf;
 
-	if (!buf) {
-		/* First run, allocate the buffer. */
-		buf = kmem_cache_alloc(sc->sc_rxbuf_cache, kmflag);
-		sc->sc_rxbufs[ve->qe_index] = buf;
-	}
+		ve = vq_alloc_entry(sc->sc_rx_vq);
+		if (!ve) {
+			/*
+			 * Out of free descriptors - ring already full.
+			 * It would be better to update sc_norxdescavail
+			 * but MAC does not ask for this info, hence we
+			 * update sc_norecvbuf.
+			 */
+			sc->sc_norecvbuf++;
+			break;
+		}
+		buf = sc->sc_rxbufs[ve->qe_index];
 
-	/* Still nothing? Bye. */
-	if (!buf) {
-		dev_err(sc->sc_dev, CE_WARN, "Can't allocate rx buffer");
-		sc->sc_norecvbuf++;
-		goto exit_buf;
-	}
+		if (!buf) {
+			/* First run, allocate the buffer. */
+			buf = kmem_cache_alloc(sc->sc_rxbuf_cache, kmflag);
+			sc->sc_rxbufs[ve->qe_index] = buf;
+		}
 
-	ASSERT(buf->rb_mapping.vbm_ncookies >= 1);
+		/* Still nothing? Bye. */
+		if (!buf) {
+			dev_err(sc->sc_dev, CE_WARN,
```

Comment:
I think it could potentially flood the console if kmem_cache_alloc() fails for some period of time and we continue to get rx interrupts during that time. Since we already have a norcvbuf kstat to diagnose such a condition, this message can probably just go away.